larsim/PhotonPropagation/PhotonLibrary.cxx

```diff
@@ -398,6 +438,127 @@ namespace phot{
     else return fReflTLookupTable.data_address(uncheckedIndex(Voxel, 0));
   }
 
+  //------------------------------------------------------------
+  void PhotonLibrary::LoadMetadata(TDirectory& srcDir) {
+    
+    constexpr std::size_t NExpectedKeys = 9U;
+    
+    std::vector<std::string> missingKeys;
+    
+    RooReader<RooInt, Int_t> readInt { srcDir, missingKeys };
+    RooReader<RooDouble, Double_t> readDouble { srcDir, missingKeys };
+    
+    double xMin;
+    double xMax;
+    int xN;
+    double yMin;
+    double yMax;
+    int yN;
+    double zMin;
+    double zMax;
+    int zN;
+    if (auto metaValue = readDouble("MinX")) xMin = *metaValue;
+    if (auto metaValue = readDouble("MaxX")) xMax = *metaValue;
+    if (auto metaValue = readInt  ("NDivX")) xN   = *metaValue;
+    if (auto metaValue = readDouble("MinY")) yMin = *metaValue;
+    if (auto metaValue = readDouble("MaxY")) yMax = *metaValue;
+    if (auto metaValue = readInt  ("NDivY")) yN   = *metaValue;
+    if (auto metaValue = readDouble("MinZ")) zMin = *metaValue;
+    if (auto metaValue = readDouble("MaxZ")) zMax = *metaValue;
+    if (auto metaValue = readInt  ("NDivZ")) zN   = *metaValue;
+    
+    if (!missingKeys.empty()) {
+      if (missingKeys.size() != NExpectedKeys) {
+        mf::LogError log("PhotonLibrary");
+        log
+          << "Photon library at '" << srcDir.GetPath() << "' is missing "
+          << missingKeys.size() << " metadata elements:";
+        for (auto const& key: missingKeys) log << " '" << key << "'";
+      }
+      else {
+        mf::LogTrace("PhotonLibrary")
+          << "No voxel metadata found in '" << srcDir.GetPath() << "'";
+      }
+      return;
+    } // if missing keys
+    
+    fVoxelDef.emplace(
+      xMin, xMax, xN,
+      yMin, yMax, yN,
+      zMin, zMax, zN
+      );
+    
+  } // PhotonLibrary::LoadMetadata()
+  
+  
+  //------------------------------------------------------------
+  void PhotonLibrary::StoreMetadata() const {
+    
+    auto const& tfs = *(art::ServiceHandle<art::TFileService const>());
```

Comment:
My preference would be for the @art::TFileService@ reference to be passed in instead of creating a handle inside a file that is not intrinsically related to art.  The pattern you've adopted is consistent with what has been done before, but LArSoft needs to be getting away from it.

Reply:
Ok, I will humour you, with the understanding that I consider `PhotonVisibilityService` a lost battle.
I have moved _art_ services out of `PhotonLibrary` entirely. It will be passed in by the caller, if the caller thinks it's needed.
And if the caller thinks wrong, they'll get an exception at the right time.
In `PhotonVisibilityService`, `TFileService` is required if building the library, optional if there is a request not to load the library, and skipped otherwise.

Reply:
Commit fcbd706.